engine/table/src/main/java/io/deephaven/engine/table/impl/UpdateByOperatorFactory.java

```diff
@@ -0,0 +1,356 @@
+                : Arrays.stream(groupByColumns)
+                        .map(MatchPair::right).collect(Collectors.toSet());
+        return table.getDefinition().getColumnStream()
+                .filter(c -> !usedGroupColumns.contains(c.getName()) && spec.applicableTo(c.getDataType()))
+                .map(c -> new MatchPair(c.getName(), c.getName()))
+                .toArray(MatchPair[]::new);
+    }
+
+    public String describe(Collection<UpdateByClause> clauses) {
+        final Describer d = new Describer();
+        clauses.forEach(c -> c.walk(d));
+        return d.descriptionBuilder.toString();
+    }
+
+    private static class Describer implements UpdateByClause.Visitor {
+        final StringBuilder descriptionBuilder = new StringBuilder();
+        String columnStr;
+
+        @Override
+        public void visit(ColumnUpdateClause clause) {
+            final MatchPair[] pairs = parseMatchPairs(clause.getColumns());
+            if (pairs.length == 0) {
+                columnStr = "[All]";
+            } else {
+                columnStr = MatchPair.matchString(pairs);
+            }
+
+            descriptionBuilder.append(clause.getSpec().describe()).append("(").append(columnStr).append("), ");
+            columnStr = null;
+        }
+    }
+
+    private class OperationVisitor implements UpdateBySpec.Visitor, UpdateByClause.Visitor {
+        private final List<UpdateByOperator> ops = new ArrayList<>();
+        private MatchPair[] pairs;
+
+        /**
+         * Check if the supplied type is one of the supported time types.
+         * 
+         * @param type the type
+         * @return true if the type is one of the useable time types
+         */
+        public boolean isTimeType(final @NotNull Class<?> type) {
+            // TODO: extend time handling similar to enterprise (Instant, ZonedDateTime, LocalDate, LocalTime)
+            return type == DateTime.class;
+        }
+
+        @Override
+        public void visit(@NotNull final ColumnUpdateClause clause) {
+            final UpdateBySpec spec = clause.getSpec();
+            pairs = createColumnsToAddIfMissing(source, parseMatchPairs(clause.getColumns()), spec, groupByColumns);
+            spec.walk(this);
+            pairs = null;
+        }
+
+        @Override
+        public void visit(@NotNull final EmaSpec ema) {
+            final String timestampCol = ema.getTimestampCol();
+            final LongRecordingUpdateByOperator timeStampRecorder;
+            final boolean isTimeBased = !StringUtils.isNullOrEmpty(timestampCol);
+
+            if (isTimeBased) {
+                timeStampRecorder = makeLongRecordingOperator(source, timestampCol);
+                ops.add(timeStampRecorder);
+            } else {
+                timeStampRecorder = null;
+            }
+
+            Arrays.stream(pairs)
+                    .filter(p -> !isTimeBased || !p.right().equals(timestampCol))
```

Comment:
This equals is between incompatible types - makes me think that ema.getTimestampCol() should be ColumnName.